File: compscieng2/compscieng_2_19/cg.py
import numpy as np
import scipy.linalg as lin

A = np.array([[6.,1.,1.],
              [1.,7.,1.],
              [1.,1.,8.]])

b = np.array([1.,1.,1.])

xreal = lin.solve(A, b); print "solution", xreal

p = b
r = b
x = b*0;
r2 = np.dot(r.T,r)
for i in range(5):
    Ap = np.dot(A,p)
    alpha = r2 / np.dot(p.T,Ap)
    x = x + np.dot(alpha,p)
    r = r-np.dot(alpha,Ap)
    r2old = r2
    r2 = np.dot(r.T,r)
    beta = r2 / r2old
    p = r + np.dot(beta,p)
    print x

       
    
    

File: compscieng2/compscieng_2_19/cg.m
A = [6 1 1; 1 7 1; 1 1 8];
b = [1;1;1]           
disp(A\b);

d = b;                             % initial search direction
r = b;                             % initial residual
x = b*0;                           % initial solution
r2 = r'*r;
disp('CG solution');
for i = 1:10
   Ad = A*d;                       % apply the matrix A
   alpha = r2/(d'*Ad);             % a first scalar product
   x = x+alpha*d;                  % update solution
   r = r-alpha*Ad;                 % update residual
   r2old = r2;
   r2 = r'*r;                      % a second scalar product
   beta = r2/r2old;               
   d = r+beta*d;                   % new search direction
   disp(x);
   disp("\n");
end

